### PR TITLE
Added Dockerfile in ihr-website

### PR DESCRIPTION
**Pull request by Srijan-SS02**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by Srijan-SS02** [comment]
@romain-fontugne please review it

**Comment by Srijan-SS02** [comment]
for #289

**Comment by anmode** [comment]
Hii! I tested it but it faild: 
```
#8 84.43 npm ERR! code 1
#8 84.43 npm ERR! path /ihr_website/node_modules/deasync
#8 84.44 npm ERR! command failed
#8 84.44 npm ERR! command sh -c -- node ./build.js
#8 84.44 npm ERR! gyp info it worked if it ends with ok
#8 84.44 npm ERR! gyp info using node-gyp@9.1.0
#8 84.44 npm ERR! gyp info using node@16.20.0 | linux | arm64
#8 84.44 npm ERR! gyp ERR! find Python 
#8 84.44 npm ERR! gyp ERR! find Python Python is not set from command line or npm configuration
#8 84.44 npm ERR! gyp ERR! find Python Python is not set from environment variable PYTHON
#8 84.44 npm ERR! gyp ERR! find Python checking if "python3" can be used
#8 84.44 npm ERR! gyp ERR! find Python - "python3" is not in PATH or produced an error
#8 84.44 npm ERR! gyp ERR! find Python checking if "python" can be used
#8 84.44 npm ERR! gyp ERR! find Python - "python" is not in PATH or produced an error
#8 84.44 npm ERR! gyp ERR! find Python 
#8 84.44 npm ERR! gyp ERR! find Python **********************************************************
#8 84.44 npm ERR! gyp ERR! find Python You need to install the latest version of Python.
#8 84.44 npm ERR! gyp ERR! find Python Node-gyp should be able to find and use Python. If not,
#8 84.44 npm ERR! gyp ERR! find Python you can try one of the following options:
#8 84.44 npm ERR! gyp ERR! find Python - Use the switch --python="/path/to/pythonexecutable"
#8 84.44 npm ERR! gyp ERR! find Python   (accepted by both node-gyp and npm)
#8 84.44 npm ERR! gyp ERR! find Python - Set the environment variable PYTHON
#8 84.44 npm ERR! gyp ERR! find Python - Set the npm configuration variable python:
#8 84.44 npm ERR! gyp ERR! find Python   npm config set python "/path/to/pythonexecutable"
#8 84.44 npm ERR! gyp ERR! find Python For more information consult the documentation at:
#8 84.44 npm ERR! gyp ERR! find Python https://github.com/nodejs/node-gyp#installation
#8 84.44 npm ERR! gyp ERR! find Python **********************************************************
#8 84.44 npm ERR! gyp ERR! find Python 
#8 84.44 npm ERR! gyp ERR! configure error 
#8 84.44 npm ERR! gyp ERR! stack Error: Could not find any Python installation to use
#8 84.44 npm ERR! gyp ERR! stack     at PythonFinder.fail (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:330:47)
#8 84.44 npm ERR! gyp ERR! stack     at PythonFinder.runChecks (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:159:21)
#8 84.44 npm ERR! gyp ERR! stack     at PythonFinder.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:202:16)
#8 84.44 npm ERR! gyp ERR! stack     at PythonFinder.execFileCallback (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:294:16)
#8 84.44 npm ERR! gyp ERR! stack     at exithandler (node:child_process:410:5)
#8 84.44 npm ERR! gyp ERR! stack     at ChildProcess.errorhandler (node:child_process:422:5)
#8 84.44 npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
#8 84.44 npm ERR! gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)
#8 84.44 npm ERR! gyp ERR! stack     at onErrorNT (node:internal/child_process:485:16)
#8 84.44 npm ERR! gyp ERR! stack     at processTicksAndRejections (node:internal/process/task_queues:83:21)
#8 84.44 npm ERR! gyp ERR! System Linux 5.15.49-linuxkit
#8 84.44 npm ERR! gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
#8 84.44 npm ERR! gyp ERR! cwd /ihr_website/node_modules/deasync
#8 84.44 npm ERR! gyp ERR! node -v v16.20.0
#8 84.44 npm ERR! gyp ERR! node-gyp -v v9.1.0
#8 84.44 npm ERR! gyp ERR! not ok 
#8 84.44 npm ERR! Build failed
#8 84.44 
#8 84.44 npm ERR! A complete log of this run can be found in:
#8 84.44 npm ERR!     /root/.npm/_logs/2023-04-07T09_50_46_668Z-debug-0.log
------
executor failed running [/bin/sh -c npm install]: exit code: 1 
```
More probably i would say for mac.

So i created this image:
```
# Use the official Node.js image
FROM node:16.9.1-alpine

# Install Python and other dependencies
RUN apk add --no-cache python3 make g++

# Set the working directory
WORKDIR /app

# Copy package.json and package-lock.json to the container
COPY package*.json ./

# Install npm packages
RUN npm install

# Copy the rest of the application code to the container
COPY . .

# Expose the port used by the application (change this if necessary)
EXPOSE 8080

# Run the application
CMD ["npm", "run", "serve"]
```
Note: instead of copying the entire content of the current directory to /ihr_website, we only copy the package.json and package-lock.json files to /app first. This allows Docker to cache the dependency installation step if these files haven't changed, resulting in faster builds.

I was trying to do a multistage building to reduce the size. I tried but there were some problem. Unfortunately My exams are there till 15 april.

**Comment by Srijan-SS02** [comment]
> 

Hey @anmode , yes that's only for mac I think. Because I'm able to build images and run container in Linux  systems though.
 
``` [ss02@192 ihr-website]$ docker build -t ihr-website:latest1 .
[+] Building 2.9s (10/10) FINISHED                                                                                                               
 => [internal] load build definition from dockerfile                                                                                        0.1s
 => => transferring dockerfile: 32B                                                                                                         0.0s
 => [internal] load .dockerignore                                                                                                           0.1s
 => => transferring context: 2B                                                                                                             0.0s
 => [internal] load metadata for docker.io/library/node:16-alpine                                                                           2.4s
 => [1/5] FROM docker.io/library/node:16-alpine@sha256:b4a72f83f52bbe3970bb74a15e44ec4cf6e873ad4787473dfc8a26f5b4e29dd2                     0.0s
 => [internal] load build context                                                                                                           0.1s
 => => transferring context: 11.86kB                                                                                                        0.0s
 => CACHED [2/5] WORKDIR /ihr_website                                                                                                       0.0s
 => CACHED [3/5] COPY . /ihr_website/                                                                                                       0.0s
 => CACHED [4/5] RUN npm install                                                                                                            0.0s
 => CACHED [5/5] RUN export NODE_OPTIONS=--openssl-legacy-provider                                                                          0.0s
 => exporting to image                                                                                                                      0.0s
 => => exporting layers                                                                                                                     0.0s
 => => writing image sha256:fa3e98380d617dc4585adf91021591300a0cdddbd02788395a5e1b00e6d6e9db                                                0.0s
 => => naming to docker.io/library/ihr-website:latest1 
```
Thanks for the input, will definitely keep that in mind and update the dockerfile, considering mac OS as well.

**Comment by anmode** [comment]
> > 
> 
> Hey @anmode , yes that's only for mac I think. Because I'm able to build images and run container in Linux  systems though.
>  
> ``` [ss02@192 ihr-website]$ docker build -t ihr-website:latest1 .
> [+] Building 2.9s (10/10) FINISHED                                                                                                               
>  => [internal] load build definition from dockerfile                                                                                        0.1s
>  => => transferring dockerfile: 32B                                                                                                         0.0s
>  => [internal] load .dockerignore                                                                                                           0.1s
>  => => transferring context: 2B                                                                                                             0.0s
>  => [internal] load metadata for docker.io/library/node:16-alpine                                                                           2.4s
>  => [1/5] FROM docker.io/library/node:16-alpine@sha256:b4a72f83f52bbe3970bb74a15e44ec4cf6e873ad4787473dfc8a26f5b4e29dd2                     0.0s
>  => [internal] load build context                                                                                                           0.1s
>  => => transferring context: 11.86kB                                                                                                        0.0s
>  => CACHED [2/5] WORKDIR /ihr_website                                                                                                       0.0s
>  => CACHED [3/5] COPY . /ihr_website/                                                                                                       0.0s
>  => CACHED [4/5] RUN npm install                                                                                                            0.0s
>  => CACHED [5/5] RUN export NODE_OPTIONS=--openssl-legacy-provider                                                                          0.0s
>  => exporting to image                                                                                                                      0.0s
>  => => exporting layers                                                                                                                     0.0s
>  => => writing image sha256:fa3e98380d617dc4585adf91021591300a0cdddbd02788395a5e1b00e6d6e9db                                                0.0s
>  => => naming to docker.io/library/ihr-website:latest1 
> ```
> Thanks for the input, will definitely keep that in mind and update the dockerfile, considering mac OS as well. 

That's cool! Btw what's the size of image and is the container serving localhost correctly.
Okay then I'll create pR for arm64 nd windows.

**Comment by Srijan-SS02** [comment]
Hey @anmode , of-course it was cached, it can be easily understood I believe. And size was around 1 GB before multi staging. So I have created a multistaged docker file for the same.
``` 
#BUILDER
FROM node:16.9.1-alpine as builder

RUN apk add --no-cache python3 make g++

#Builder directory
WORKDIR /ihr-website

#Copy builder source
COPY package*.json ./

# installing all dependencies
RUN npm install

# Copy all files
COPY . .

# MAIN
FROM node:16.9.1-alpine as main

# Make main app directory
WORKDIR /app

#Copy all from builder stage
COPY --from=builder /ihr-website/.  ./

#Start the serve
CMD ["npm", "run", "serve"]
```
Size after muti-stage building is 599 MB.
But will still leave upon @romain-fontugne what he thinks about this, before I edit my PR. 
Thanks for your valuable input @anmode :smiley:

**Comment by romain-fontugne** [comment]
I have no idea :D
what is the purpose of the multi-stage? is it only to save space?

**Comment by Srijan-SS02** [comment]
@romain-fontugne  technically yes, decreasing images size is one of the feature. Some other features are
-  it is faster when we push this image to   some container registry. 
-  Reduces CI/CD job time while using smaller docker image
- Some other feature on this [Link](https://medium.com/@chzbrgr71/docker-multi-stage-builds-size-matters-e71ae0f08ae8) 


Moreover could be called good docker practice.

**Comment by anmode** [comment]
> @romain-fontugne  technically yes, decreasing images size is one of the feature. Some other features are
> -  it is faster when we push this image to   some container registry. 
> -  Reduces CI/CD job time while using smaller docker image
> - Some other feature on this [Link](https://medium.com/@chzbrgr71/docker-multi-stage-builds-size-matters-e71ae0f08ae8) 
> 
> 
> Moreover could be called good docker practice.

Yeppp...and we can decrease image size to 150mb in this 😂
Actually alpine is larger we have other options too and through multiple stages we can reduce further.

**Comment by romain-fontugne** [comment]
OK, thanks, let's do multi-stage then!

File: dockerfile [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#BUILDER
+FROM node:16.9.1-alpine as builder
+
+RUN apk add --no-cache python3 make g++
```
Comment (anmode) [inline dockerfile:4]: I think I used it for Mac! as my machine needs it. You can clean it when creating the main directory.(so size will be reduced as each mb matters). Btw just a question I hope this image is working properly in Windows/Linux.

File: dockerfile [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+COPY package*.json ./
+
+# installing all dependencies
+RUN npm install
+
+# Copy all files
+COPY . .
+
+# MAIN
+FROM node:16.9.1-alpine as main
```
Comment (anmode) [inline dockerfile:19]: We can use Slim...as it smallest and in our case would be a good option. https://github.com/slimtoolkit/slim

**Review by anmode — COMMENTED** [review]
These are just suggestions!! Btw nice we are able to reduce the image to 580MB.

File: dockerfile (hunk 1 above)
Comment (Srijan-SS02) [reply dockerfile:4]: Yeaah this is specifically for M1 only but decided to continue with it for M1 and fedora users else we need to mention this in README.

File: dockerfile (hunk 2 above)
Comment (Srijan-SS02) [reply dockerfile:19]: Actually I am trying to update the dockerfile with more stages will comment it in few hours

Comment (Srijan-SS02) [reply dockerfile:19]: And definitely we can try with slim as well. :smile:

Comment (Srijan-SS02) [reply dockerfile:19]: Slim isn't helping ...infact when I tried size got increased some how

Comment (anmode) [reply dockerfile:19]: > Slim isn't helping ...infact when I tried size got increased some how 

Idk as i far i know...this should not happen.
Could you check the layers?
As there would be three images. One is alpine ,slim nd our ihr
 And content may be something is creating problem. 
If don't wanna go in depth...it is ok alpine.
I would see and tried on local machine afterwards 😅

Comment (Srijan-SS02) [reply dockerfile:19]: No, not a problem. I'll check the layers and let you know

**Comment by Srijan-SS02** [comment]
I have edited the dockerfile to 3 stages to optimize it and the size is reduced to almost 80mb

```
FROM node:16.9.1-alpine as dependencies
WORKDIR /app
RUN apk add --no-cache python3 make g++
COPY package*.json ./
RUN npm install


FROM node:16.9.1-alpine as build
WORKDIR /app
COPY . .
COPY --from=dependencies /app/node_modules ./node_modules
RUN npm run build


FROM nginx:alpine
EXPOSE 80
COPY --from=build /app/dist /usr/share/nginx/html
```
I
but when I'm opening this on my localhost, it only opens till here.

Screenshot from 2023-04-22 23-42-11


@romain-fontugne  @anmode  suggest anything if that can help

**Comment by anmode** [comment]
> @romain-fontugne @anmode suggest anything if that can help

ahh! great. You did nicely multi-stage build.
so after going into the docker container I found what causing the problem.

see  Image structure:
Screenshot 2023-04-23 at 3 51 33 PM

Here is error messages:
Screenshot 2023-04-23 at 4 00 53 PM

```
"/usr/share/nginx/html/ihr/en-us/index.html" is not found (2: No such file or directory), client: 172.17.0.1, server: localhost, request: "GET /ihr/en-us/ HTTP/1.1", host: "localhost"
```



Based on the error message, it seems that Nginx is trying to find the index.html file in the `/usr/share/nginx/html/ihr/en-us/ `
directory, but index.html is present in `/usr/share/nginx/html/` -refer the pic i have attached. Try to solve needs a minor changes ig. I hope it helps.

**Comment by Srijan-SS02** [comment]
> > @romain-fontugne @anmode suggest anything if that can help
> 
> ahh! great. You did nicely multi-stage build. so after going into the docker container I found what causing the problem.
> 
> see Image structure: Screenshot 2023-04-23 at 3 51 33 PM
> 
> Here is error messages: Screenshot 2023-04-23 at 4 00 53 PM
> 
> ```
> "/usr/share/nginx/html/ihr/en-us/index.html" is not found (2: No such file or directory), client: 172.17.0.1, server: localhost, request: "GET /ihr/en-us/ HTTP/1.1", host: "localhost"
> ```
> 
> Based on the error message, it seems that Nginx is trying to find the index.html file in the `/usr/share/nginx/html/ihr/en-us/ ` directory, but index.html is present in `/usr/share/nginx/html/` -refer the pic i have attached. It needs minor changes. The game is behind just copying the files so that when they are requested they can be located and loaded.

Thanks @anmode.
As far as I can understand, I should copy then content where it is looking for the index.html that is over here `/usr/share/nginx/html/ihr/en-us/ `.
 But when I  am doing this `COPY --from=build /app/dist /usr/share/nginx/html/ihr/en-us/` it is displaying the default page of nginx.
Am I doing anything wrong??

**Comment by anmode** [comment]
Oh really so sorry for some days. I'm away from my machine.

 I think you missed something....please could you check the image layer in container see where the files are being copied.
It will show error that dir not found..so it would help to debug. 
I'll share some docs.. for now really sorry.

**Comment by Srijan-SS02** [comment]
No that's not a problem even I took one. sorry for late reply too

Well here are the screen-shots of the content when I have changed the path

for `/usr/share/nginx/html/`
Screenshot from 2023-04-29 11-58-02
Here the same error continues

And when path is `/usr/share/nginx/html/ihr/en-us/` then
Screenshot from 2023-04-29 12-13-13
but in this case it is loading nginx default html page.

And what I found is that vue.js file has path defined as ` publicPath: '/ihr/',`  instead of  `" "`see here in last line 
Screenshot from 2023-04-29 12-17-24

And ihr is using reverse proxy  so we cannot change the path untill mentors allow to do so.

So every time we use this path `/usr/share/nginx/html/` which is where files should actually be because nginx.conf looks them for there, @anmode  check in nginx.conf file inside container. Browsers directs them to /ihr/en-us and we se nothing but error. Only way to solve it that I could have understand  is by changing nginx.conf file. 
see here
Screenshot from 2023-04-29 12-40-05

What are your views on it @anmode @romain-fontugne 
shall we continue with previous dockerfile with size of about 500 mb and later we can resolve it.

**Comment by Srijan-SS02** [comment]
@romain-fontugne  @anmode I have made changes to the dockerfile and its working good on my machine @romain-fontugne please review it.
Commits have been littles messy if you want i can create a new fork and then put another PR or we can continue with the same. Once you can tell everything working great I will make changes in README file for docker as well.
```
FROM node:16.9.1-alpine as dependencies
WORKDIR /app
RUN apk add --no-cache python3 make g++
COPY package*.json ./
RUN npm install


FROM node:16.9.1-alpine as build
WORKDIR /app
COPY . .
RUN sed -i '/publicPath:/d' vue.config.js
COPY --from=dependencies /app/node_modules ./node_modules
RUN npm run build

FROM nginx:1.20-alpine
COPY --from=build /app/dist /usr/share/nginx/html/

```

size of docker file is now `57.3 MB`

**Comment by dpgiakatos** [comment]
The application has been upgraded to Vue 3.js with a new file structure. Kindly acquaint yourself with the updated code structure before revisiting this pull request.

**Comment by dpgiakatos** [comment]
Hello, since this PR has been open for an extended period and hasn't been updated to align with our new Vue3JS code structure, we've addressed its changes in PR #719. Additionally, the new size of the Docker image is 25.41MB.

**Pull request closed without merging**